app/src/main/java/org/wildaid/ofish/ui/search/complex/ComplexSearchViewModel.kt

```diff
@@ -69,96 +90,113 @@ class ComplexSearchViewModel(repository: Repository, application: Application) :
         private val addSearchModel = AddSearchModel(R.string.add_new_vessel)
         private var cachedAllReports = emptyList<Report>()
 
-        override fun initiateData(): List<SearchModel> {
-            fetchReports()
-
-            val result = mutableListOf<SearchModel>()
-            if (isAddAvailable) result.add(addSearchModel)
-            if (cachedAllReports.isNotEmpty()) result.add(TextViewSearchModel(R.string.recently_boarded))
-
-            result.addAll(cachedAllReports
-                .asSequence()
-                .filterNot { it.vessel?.name.isNullOrBlank() }
-                .groupBy { it.vessel?.permitNumber to it.vessel?.name }
-                .map { pair ->
-                    RecordSearchModel(
-                        pair.value.find { (it.vessel?.permitNumber to it.vessel?.name) == pair.key }?.vessel!!,
-                        pair.value.sortedByDescending { report -> report.date }, repository
-                    )
-                }.take(RECENT_BOARDINGS_COUNT)
-            )
-            return result
+        override fun initiateData(): Flow<List<SearchModel>> {
+            return repository.findReportsGroupedByVessel()
+                .map { reports ->
+                    cachedAllReports = reports
+
+                    mutableListOf<SearchModel>().apply {
+                        if (isAddAvailable) {
+                            this.add(addSearchModel)
+                        }
+                        if (reports.isNotEmpty()) {
+                            this.add(TextViewSearchModel(R.string.recently_boarded))
+                        }
+
+                        this.addAll(
+                            reports.asSequence()
+                                .filterNot { it.vessel?.name.isNullOrBlank() }
+                                .groupBy { it.vessel?.permitNumber to it.vessel?.name }
+                                .map { pair ->
+                                    RecordSearchModel(
+                                        pair.value.find { (it.vessel?.permitNumber to it.vessel?.name) == pair.key }?.vessel!!,
```

Comment:
I know this is the original code, but using `it.vessel?.permitNumber` here means that `null` might be used as a key. I would probably categorize it as a bug, but 🤷